modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/block/BlockStorage.scala

```diff
@@ -22,199 +21,116 @@ import org.tessellation.security.signature.Signed
+            .map(Tips(_))
         case _ => None
       }
-    } yield pulledTips
 
   private def isBlockAccepted(blockReference: BlockReference): F[Boolean] =
-    blocks(blockReference.hash).get.map(_.exists(_.isInstanceOf[AcceptedBlock]))
-
-  private def updateNoTipParents(parents: List[ProofsHash], hashedBlock: Hashed[DAGBlock]): F[Unit] =
-    parents.traverse { hash =>
-      blocks(hash)
-        .modify[Either[TipsUpdateError, Unit]] {
-          case Some(accepted @ AcceptedBlock(_, _, _, false)) =>
-            (accepted.addUsage(hashedBlock.proofsHash).some, ().asRight)
-          case unexpected =>
-            (unexpected, NoTipAcceptedBlockNotFound(hash, unexpected).asLeft)
-        }
-        .flatMap(_.liftTo[F])
-    }.void
-
-  private def updateTipParentsToRemove(parents: List[ProofsHash], hashedBlock: Hashed[DAGBlock]): F[Unit] =
-    parents.traverse { hash =>
-      blocks(hash)
-        .modify[Either[TipsUpdateError, Unit]] {
-          case Some(accepted @ AcceptedBlock(_, _, _, true)) =>
-            (accepted.addUsage(hashedBlock.proofsHash).unsetAsTip.some, ().asRight)
-          case unexpected =>
-            (unexpected, TipBlockToRemoveAsTipNotFound(hash, unexpected).asLeft)
-        }
-        .flatMap(_.liftTo[F])
-    }.void
-
-  private def updateTipParentsToRemain(parents: List[ProofsHash], hashedBlock: Hashed[DAGBlock]): F[Unit] =
-    parents.traverse { hash =>
-      blocks(hash)
-        .modify[Either[TipsUpdateError, Unit]] {
-          case Some(accepted @ AcceptedBlock(_, _, _, true)) =>
-            (accepted.addUsage(hashedBlock.proofsHash).some, ().asRight)
-          case unexpected =>
-            (unexpected, TipBlockToRemainAsTipNotFound(hash, unexpected).asLeft)
-        }
-        .flatMap(_.liftTo[F])
-    }.void
-
-  private def tryAddAsATip(condition: Boolean, hashedBlock: Hashed[DAGBlock]): F[Unit] =
-    if (condition)
-      blocks(hashedBlock.proofsHash).modify {
-        case Some(accepted @ AcceptedBlock(_, _, usages, false)) if usages.size < tipsConfig.maximumTipUsages =>
-          (accepted.setAsTip.some, ())
-        case other => (other, ())
-      } else Applicative[F].unit
+    blocks(blockReference.hash).get.map(_.exists(_.isInstanceOf[MajorityBlock]))
 
   private def updateParentUsages(hashedBlock: Hashed[DAGBlock]): F[Unit] = {
     val hash = hashedBlock.proofsHash
     hashedBlock.signed.value.parent.toList.traverse { blockReference =>
-      blocks(blockReference.hash).update {
+      blocks(blockReference.hash).modify {
         case Some(storedBlock) =>
           storedBlock match {
-            case unknown: UnknownBlock   => unknown.addUsage(hash).some
-            case waiting: WaitingBlock   => waiting.addUsage(hash).some
-            case accepted: AcceptedBlock => accepted.addUsage(hash).some
+            case waiting: WaitingBlock =>
+              (waiting.some, TipUsageUpdateError(hash, blockReference.hash, "Waiting").asLeft)
```

Comment:
I see "Waiting", "Accepted", "Majority", .. a lot here. Would you mind creating some enum or more specific errors?

Reply:
I purposefully flattened multiple errors (that we had before) into one more general with this additional field indicating what unexpected state was encountered. We have one error instead of three. I can go to the previous concept, but I remember that some time ago we discussed that single error per case may be overkill, so I changed it here. I was thinking about packing the encountered state (`Option[StoredBlock]`) into the error message and then dynamically creating the message, but I guess we also don't want to carry the whole object with state inside of an error class object and not all encountered states are erroneous. I could create an enum for this states (Waiting, Accepted, Majority, NotFound) but then they reflect the `StoredBlock` abstraction mostly and at the same time they are decoupled from that abstraction. In the end I decided that maybe just using String is good enough. We don't match these errors anywhere currently, they are only for information purposes. If you have some idea on how to efficiently tackle it without creating too much boilerplate, please share.

Reply:
You can use a sealed trait for each error and then specify more specific case classes. Also, if you would want to have a block inside the error I'd go with a companion object that does the logic and keep an error class simple. Companion object "apply" will use the full object just for creating the error rather than storing the object all the time.

Reply:
Having unique errors for each kind of block would be helpful if handling differs depending on the type. Otherwise parametrized error IMO is ok.

Reply:
Ok I discussed it with @tbekas and we have come to a conclusion that by using `implicit show: Show[StoredBlock]` we can achieve the precise messages with no surplus boilerplate. We don't think there is any penalty in putting the whole `Option[StoredBlock]` into the error message as it's anyway just a reference to an object already in memory.